matrix_sdk/src/client.rs

```diff
@@ -692,8 +738,244 @@ impl Client {
+    ///             // Open sso_url
+    ///             Ok(())
+    ///         },
+    ///         None,
+    ///         None,
+    ///         None,
+    ///         Some("My app")
+    ///     )
+    ///     .await
+    ///     .unwrap();
+    ///
+    /// println!("Logged in as {}, got device_id {} and access_token {}",
+    ///          response.user_id, response.device_id, response.access_token);
+    /// # })
+    /// ```
+    ///
+    /// [`get_sso_login_url`]: #method.get_sso_login_url
+    /// [`login_with_token`]: #method.login_with_token
+    /// [`restore_login`]: #method.restore_login
+    #[cfg(feature = "sso_login")]
+    #[cfg_attr(feature = "docs", doc(cfg(sso_login)))]
+    pub async fn login_with_sso<C>(
+        &self,
+        use_sso_login_url: impl Fn(String) -> C,
+        server_url: Option<&str>,
+        server_response: Option<&str>,
+        device_id: Option<&str>,
+        initial_device_display_name: Option<&str>,
+    ) -> Result<login::Response>
+    where
+        C: Future<Output = Result<()>>,
+    {
+        info!("Logging in to {}", self.homeserver);
+        let (signal_tx, signal_rx) = oneshot::channel();
+        let (data_tx, data_rx) = oneshot::channel();
+        let data_tx_mutex = Arc::new(std::sync::Mutex::new(Some(data_tx)));
+
+        let mut redirect_url = match server_url {
+            Some(s) => match Url::parse(s) {
+                Ok(url) => url,
+                Err(err) => return Err(IoError::new(IoErrorKind::InvalidData, err).into()),
+            },
+            None => Url::parse("http://localhost:0/").unwrap(),
+        };
+
+        let response = match server_response {
+            Some(s) => s.to_string(),
+            None => String::from("<h2>The Single Sign-On login process is complete.</h2><p>You can close this page now.</p>"),
+        };
+
+        let route = warp::get()
+            .and(warp::query::<HashMap<String, String>>())
+            .map(move |p: HashMap<String, String>| {
+                if let Some(data_tx) = data_tx_mutex.lock().unwrap().take() {
+                    if let Some(token) = p.get("loginToken") {
+                        data_tx.send(Some(token.to_owned())).unwrap();
+                    } else {
+                        data_tx.send(None).unwrap();
+                    }
+                }
+                Response::builder().body(response.clone())
+            });
+
+        let redirect_addr = (
+            redirect_url.host_str().unwrap(),
+            redirect_url.port().unwrap(),
+        )
+            .to_socket_addrs()
+            .expect("Could not convert server_url to ToSocketAddrs")
+            .next()
+            .expect("Could not convert server_url to SocketAddr");
+
+        let (addr, server) = warp::serve(route).bind_with_graceful_shutdown(redirect_addr, async {
+            signal_rx.await.ok();
+        });
+        redirect_url.set_port(Some(addr.port())).unwrap();
+
+        tokio::spawn(server);
```

Comment:
I agree that in WASM using the server doesn't make sense. I should add `#[cfg(not(target_arch = "wasm32"))]` then?

Reply:
Yeah and a doc attribute so that it's clear from the docs that this doesn't work under WASM:

```rust
#[cfg_attr(
    feature = "docs",
    doc(cfg(all(sso_login, not(target_arch = "wasm32"))))
)]
```

Reply:
Should I add a compile error when the `sso_login` feature is enabled under WASM?

Reply:
Oh nice, yes that sounds great as well.